allensdk/test/brain_observatory/behavior/behavior_project_cache/utils.py

```diff
@@ -0,0 +1,154 @@
+
+    metadata_blobs: Union[dict, None]
+        A dict mapping metadata filename to bytes in the file
+
+    manifest_version: str
+        The version of the manifest (manifest will be
+        uploaded to moto3 as manifest_{manifest_version}.json)
+
+    bucket_name: str
+
+    project_name: str
+
+    client: boto3.client
+
+    Returns
+    -------
+    None
+        Uploads the provided data, generates the manifest,
+        and uploads the manifest to moto3
+    """
+
+    for fname in data_blobs:
+        client.put_object(Bucket=bucket_name,
+                          Key=f'{project_name}/data/{fname}',
+                          Body=data_blobs[fname]['data'])
+
+    if metadata_blobs is not None:
+        for fname in metadata_blobs:
+            client.put_object(Bucket=bucket_name,
+                              Key=f'{project_name}/project_metadata/{fname}',
+                              Body=metadata_blobs[fname])
+
+    response = client.list_object_versions(Bucket=bucket_name)
+    fname_to_version = {}
+    for obj in response['Versions']:
+        if obj['IsLatest']:
+            fname = obj['Key'].split('/')[-1]
+            fname_to_version[fname] = obj['VersionId']
+
+    manifest = {}
+    manifest['manifest_version'] = manifest_version
+    manifest['project_name'] = project_name
+    manifest['metadata_file_id_column_name'] = 'file_id'
+    manifest['metadata_files'] = {}
+    manifest['data_pipeline'] = [{'name': 'AllenSDK', 'version': '1.1.1'}]
+
+    data_file_dict = {}
+    url_root = f'http://{bucket_name}.s3.amazonaws.com/{project_name}/data'
+    for fname in data_blobs:
+        url = f'{url_root}/{fname}'
+        hasher = hashlib.blake2b()
+        hasher.update(data_blobs[fname]['data'])
+        checksum = hasher.hexdigest()
+
+        data_file = {'url': url,
+                     'version_id': fname_to_version[fname],
+                     'file_hash': checksum}
+
+        data_file_dict[data_blobs[fname]['file_id']] = data_file
+
+    manifest['data_files'] = data_file_dict
+
+    if metadata_blobs is not None:
+        url_root = f'http://{bucket_name}.s3.amazonaws.com/{project_name}/'
+        url_root += 'project_metadata'
+
+        metadata_dict = {}
+        for fname in metadata_blobs:
+            url = f'{url_root}/{fname}'
+            hasher = hashlib.blake2b()
+            hasher.update(metadata_blobs[fname])
+            metadata_dict[fname] = {'url': url,
+                                    'file_hash': hasher.hexdigest(),
+                                    'version_id': fname_to_version[fname]}
+
+            manifest['metadata_files'] = metadata_dict
+
+    manifest_k = f'{project_name}/manifests/'
+    manifest_k += f'{project_name}_manifest_v{manifest_version}.json'
+    client.put_object(Bucket=bucket_name,
+                      Key=manifest_k,
+                      Body=bytes(json.dumps(manifest), 'utf-8'))
+
+    return None
+
+
+def create_bucket(test_bucket_name: str,
```

Comment:
Functions like this should probably be relegated to `pytest.fixtures` since we probably don't want to give the appearance that you can create buckets and add files to them with the AllenSDK, when in fact these functions are only supposed to be used for tests.

Reply:
I would like to push back on this. The module is buried pretty deep in `test/`. If a user knows enough to find the module, they probably know enough to acknowledge what that means. Based on my understanding of pytest fixtures, it's going to take a lot of reconfiguring to convert these methods into fixtures given how I am using them, and I would rather devote that time elsewhere.

Would your fears be allayed if I added an explicit "THIS IS ONLY FOR UNIT TESTING" to the docstrings for these methods?

Reply:
I agree with Nick, this would be more in-line with pytest paradigms. On the other hand, I think this could be modified later, without any impact to the current schedule or functionality.
I propose a compromise to accept as-is and write a backlog ticket to make these fixtures.